src/SA/ldapsecuritycheck/entry.c

```diff
@@ -0,0 +1,289 @@
+{
+	return TRUE;
+}
+
+BOOL checkLDAP(wchar_t* dc, wchar_t* spn, BOOL ssl)
+{
+	CredHandle hCredential;
+	TimeStamp tsExpiry;
+	SECURITY_STATUS getHandle = SECUR32$AcquireCredentialsHandleW(
+		NULL,
+		L"NTLM",
+		SECPKG_CRED_OUTBOUND,
+		NULL,
+		NULL,
+		NULL,
+		NULL,
+		&hCredential,
+		&tsExpiry
+	);
+
+	if (getHandle != SEC_E_OK)
+	{
+		BeaconPrintf(CALLBACK_ERROR, "[-] AcquireCredentialsHandleW failed: %lu\n", getHandle);
+		return FALSE;
+	}
+
+	//ldap
+	ULONG result;
+	LDAP* pLdapConnection = NULL;
+
+	if(ssl == TRUE){
+		pLdapConnection = WLDAP32$ldap_initW(dc, 636);
+	}else{
+		pLdapConnection = WLDAP32$ldap_initW(dc, 389);
+	}
+	if (pLdapConnection == NULL)
+	{
+		BeaconPrintf(CALLBACK_ERROR, "[-] Failed to establish LDAP connection");
+		return FALSE;
+	}
+
+	const int version = LDAP_VERSION3;
+	result = WLDAP32$ldap_set_optionW(pLdapConnection, LDAP_OPT_VERSION, (void*)&version);
+
+	if(ssl == TRUE){
+        WLDAP32$ldap_get_optionW(pLdapConnection, LDAP_OPT_SSL, &result);  //LDAP_OPT_SSL
+        if (result == 0)
+            WLDAP32$ldap_set_optionW(pLdapConnection, LDAP_OPT_SSL, LDAP_OPT_ON);
+
+        WLDAP32$ldap_get_optionW(pLdapConnection, LDAP_OPT_SIGN, &result);  //LDAP_OPT_SIGN
+        if (result == 0)
+            WLDAP32$ldap_set_optionW(pLdapConnection, LDAP_OPT_SIGN, LDAP_OPT_ON);
+
+        WLDAP32$ldap_get_optionW(pLdapConnection, LDAP_OPT_ENCRYPT, &result);  //LDAP_OPT_ENCRYPT
+        if (result == 0)
+            WLDAP32$ldap_set_optionW(pLdapConnection, LDAP_OPT_ENCRYPT, LDAP_OPT_ON);
+
+        WLDAP32$ldap_set_optionW(pLdapConnection, LDAP_OPT_SERVER_CERTIFICATE, (void*)&ServerCertCallback ); //LDAP_OPT_SERVER_CERTIFICATE
+	}
+
+	result = WLDAP32$ldap_connect(pLdapConnection, NULL);
+	if (result != LDAP_SUCCESS)
+	{
+		BeaconPrintf(CALLBACK_ERROR, "[-] ldap_connect failed:");
+		return FALSE;
+	}
+
+	ULONG res;
+	struct berval* servresp = NULL;
+
+	SecBufferDesc InBuffDesc;
+	SecBuffer InSecBuff;
+
+	SECURITY_STATUS initSecurity;
+	CtxtHandle newContext;
+
+	SecBuffer secbufPointer = { 0, SECBUFFER_TOKEN, NULL };
+	SecBufferDesc output = { SECBUFFER_VERSION, 1, &secbufPointer };
+
+	SecBuffer secbufPointer3 = { 0, SECBUFFER_TOKEN, NULL };
+	SecBufferDesc output2 = { SECBUFFER_VERSION, 1, &secbufPointer };
+
+	ULONG contextAttr;
+	TimeStamp expiry;
+
+	PSecBuffer ticket;
+	int count = 0;
+	//loop
+	do {
+		if(count > 5){
+			BeaconPrintf(CALLBACK_ERROR, "[-] stuck in loop");
+			break;
+		}
+		count++;
+		if (servresp == NULL) {
+			initSecurity = SECUR32$InitializeSecurityContextW(
+				&hCredential,
+				NULL,
+				(SEC_WCHAR*)spn,
+				ISC_REQ_ALLOCATE_MEMORY | ISC_REQ_MUTUAL_AUTH | ISC_REQ_DELEGATE,
+				0,
+				SECURITY_NATIVE_DREP,
+				NULL,
+				0,
+				&newContext,
+				&output,
+				&contextAttr,
+				&expiry);
+
+			ticket = output.pBuffers;
+			//BeaconPrintf(CALLBACK_OUTPUT, "[-] size : %d\n", (DWORD)ticket->cbBuffer);
+
+			if (ticket->pvBuffer == NULL) {
```

Comment:
is output.pBuffer guaranteed to be allocated even if InitializeSecurityContext fails?
I don't think it is, meaning this could cause a crash, ticket should be checked that it isn't null prior to dereferencing it.